polys: Refactoring PolyRing and PolyElement for subsequent FLINT-Backend Integration

aasim-maverick opened the pull request:
`PolyRing` and `PolyElement` classes are refactored to separate interface and implementation specific logic to make it clearer what functionalities will be overridden upon the `FLINT` backend integration and which ones will be kept in the base class itself.

For more context see the discussion in #28106. Specifically, see: 

- Oscar Benjamin's [comment](https://github.com/sympy/sympy/issues/28106#issuecomment-2939809948) wrt [PR-28110](https://github.com/sympy/sympy/pull/28110).
- My [comment](https://github.com/sympy/sympy/issues/28106#issuecomment-2940596323) following the above. 

#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers. Formerly, `log(-x)` incorrectly gave `-log(x)`.

* physics.units
  * Corrected a semantical error in the conversion between volt and statvolt which
    reported the volt as being larger than the statvolt.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
NO ENTRY
<!-- END RELEASE NOTES -->

aasim-maverick committed:
polys: Refactoring PolyRing and PolyElement for subsequent FLINT-Backen integration

PolyRing and PolyElement classes are refactored to separate interface and
implementation specific logic to make it clearer what functionalities
will be overridden upon the FLINT backend integration and which ones will
be based on the base class itself.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -914,47 +838,40 @@ def is_nonpositive(self):
-    def is_irreducible(f):
-        if not f.ring.ngens:
-            return True
-        return f.ring.dmp_irreducible_p(f)
+    def is_irreducible(self):
+        return self._is_irreducible()
```
What is the purpose of separating `is_irreducible` from `_is_irreducible` here?

If one just calls the other then I don't see what the idea is.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1103,101 +973,15 @@ def __rsub__(p1, n):
-                    p[exp1] = v
-            return p
+            p = p1._negate()
+            return p._add_ground(n)
```
It would be better to have an `_rsub_ground` method for this. It can be done in one step with python-flint.

oscarbenjamin replied:
In python-flint this is just `n - p`:
```python
In [2]: ctx = flint.fmpz_mpoly_ctx.get(['x', 'y'])

In [3]: x, y = ctx.gens()

In [4]: c = flint.fmpz(2)

In [5]: c - x
Out[5]: -x + 2
```
We don't need to create an intermediate negated polynomial.

aasim-maverick committed:
Redundant rearrangement of property methods reversed,
_rsub_ground introduced to used in __rsub__,
_pow_int introduced to handle integer logic in __pow__ separately,
rearrangement of degree() and degrees() reversed to let subclass define them

aasim-maverick committed:
PolyRing refactored/rearranged

aasim-maverick commented:
I couldn't find much rearrangement to do for `PolyRing`, although my take at it was pretty quick, I would rather wait for a review.

aasim-maverick committed:
Refactored the previously missed PolyElement.symmetrize() method

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1227,76 +986,9 @@ def __pow__(self, n):
-            elif monom in poly:
-                del poly[monom]
+            return self._pow_single_term(n)
 
-        return poly
+        return self._pow_int(n)
```
I don't think we need to handle special cases here like `_pow_single_term`. It can just be `_pow_int` with the clear intention that `n` should be a nonnegative integer and then `_pow_int` can handle special cases. When using python-flint I don't think there is a need to handle special cases here.

aasim-maverick commented:
> I'm not sure this achieves a clean useful separation between what will happen in the superclass and what will happen in the subclass yet. The code generally needs a bit of cleanup before we can add these subclasses because they will complicate everything.

I believe I agree. What I mean to say is, that I've tried to refactor both these classes, and even though I've separated the complex interfacing logic and the implementation specific logic method by method, I've been thinking that things are still noticeably awkward. I thought I'd resolve that by back and forth review. 

My approach towards this was pretty algorithmic actually, maybe too algorithmic:

1. If a method had no implementation-specific logic, I kept it above the divider in the base class.
 
2. If a method was entirely implementation-specific, I placed it below the divider to be overridden by subclasses.
 
3. If a method had both interface and implementation-specific logic I split it into an interface method (above the divider) and implementation helpers (below the divider).

By "divider" here I mean the line I set for my convenience to set the boundary for the "above" and "below" of the class:

```
 #<--------------------------INTERFACE / IMPLEMENTATION DIVIDER-------------------------->#
```

I get an idea but I'm not exactly sure what is really meant by "a bit of cleanup". 

How should I alter my approach here ? What should I do and change about this to get it to a cleaner version of what it currently is ?

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -266,13 +266,10 @@ def __new__(cls, symbols, domain, order=lex):
-            _gens.append(poly)
+            generator = self._construct_generator(i)
+            _gens.append(generator)
         return tuple(_gens)
```
I don't think we need to change this method because the Flint version can call `.gens()` on the flint ring.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -366,47 +353,21 @@ def ring_new(self, element):
-
-        return poly
+                poly_dict[monom] = coeff
+        return self._construct_from_dict(poly_dict)
```
Maybe it would be better to call this `_from_dict_ground`.

aasim-maverick committed:
Re-refactored PolyRing

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -390,101 +428,85 @@ def _rebuild(expr):
+                raise ValueError(f"invalid generator: {gen}")
 
-        return i
+    def add_gens(self, symbols):
+        """Add new generators to the ring."""
+        syms = set(self.symbols).union(set(symbols))
+        return self.clone(symbols=list(syms))
```
`FlintPolyRing` will have a `flint_ctx` attribute which will store the context objects from `python-flint` i.e. `fmpz_mpoly_ctx` or `fmpq_mpoly_ctx`, so maybe we can use the `append_gens()` method here from python-flint here.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -582,6 +576,79 @@ def symmetric_poly(self, n):
+
+        return tuple(generators)
+
+    def _ring_equality(self, other):
+        # Check equality of two polynomial rings.
+        return ((self.symbols, self.domain, self.ngens, self.order) ==
+                (other.symbols, other.domain, other.ngens, other.order))
```
I guess in case of `python-flint` we will be able to just return `self.flint_ctx == other.flint_ctx` and that will be enough.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -582,6 +576,79 @@ def symmetric_poly(self, n):
+        return ((self.symbols, self.domain, self.ngens, self.order) ==
+                (other.symbols, other.domain, other.ngens, other.order))
+
+    def _from_dict_ground(self, element):
+        # Create polynomial from dictionary with ground domain conversion.
+        poly = self.zero
+        domain_new = self.domain_new
+
+        for monom, coeff in element.items():
+            if coeff:  # Skip zero coefficients
+                coeff = domain_new(coeff)
+                poly[monom] = coeff
+
+        return poly
```
I think we can use `self.flint_ctx.from_dict` here during override for the `python-flint` version.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -390,101 +428,85 @@ def _rebuild(expr):
+                f"got {gen}"
+            )
+
+    def _gen_index(self, gen):
+        # Get generator index from int or string.
```
For python-flint we can use `self.flint_ctx.variable_to_index()`

aasim-maverick committed:
reverting domain speicfic operations in PolyRing.from_dict()

aasim-maverick committed:
added missing domain parameter to domain_new call in _from_dict_ground

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -582,6 +576,79 @@ def symmetric_poly(self, n):
+        return ((self.symbols, self.domain, self.ngens, self.order) ==
+                (other.symbols, other.domain, other.ngens, other.order))
+
+    def _from_dict_ground(self, element, orig_domain=None):
+        # Create polynomial from dictionary with ground domain conversion.
+        poly = self.zero
+        domain_new = self.domain_new
+
+        for monom, coeff in element.items():
+            if coeff:  # Skip zero coefficients
+                coeff = domain_new(coeff, orig_domain)
+                poly[monom] = coeff
+
+        return poly
```
Exactly, I missed mentioning that in my previous comment but this issue should not persist with `python-flint`.

aasim-maverick commented:
In the most recent discussion we talked about how `PolyElement.clear_denoms` will have completely separate implementations for the pure Python and python-flint versions although while retouching the refactor of `PolyElement` I don't really get why is that necessary, since python-flint doesn't provide a separate efficient way to achieve this functionality **_yet_**.

This is how it looks like:
```python
    def clear_denoms(self):
        # In my opinion this can be split, in the meeting we decided on a complete override
        domain = self.ring.domain

        if not domain.is_Field or not domain.has_assoc_Ring:
            return domain.one, self

        ground_ring = domain.get_ring()
        common = ground_ring.one
        lcm = ground_ring.lcm
        denom = domain.denom

        for coeff in self.values():
            common = lcm(common, denom(coeff))

        poly = self.new([ (k, v*common) for k, v in self.items() ])
        return common, poly
```

And as I mentioned, I couldn't find a separate efficient way to do this in `python-flint` since a method for this functionality isn't exposed **_yet_**. So this is how I implemented it in my prototype:

```python
def itercoeffs(self):
        return iter(self.flint_poly.coeffs())

def clear_denoms(self):
        domain = self.ring.domain

        if not domain.is_Field or not domain.has_assoc_Ring:
            return domain.one, self

        ground_ring = domain.get_ring()
        common = ground_ring.one
        lcm = ground_ring.lcm
        denom = domain.denom

        for coeff in self.itercoeffs(): 
            common = lcm(common, denom(coeff))

        scaled_poly = self.flint_poly * common
        result_poly = self.ring.ring_new(scaled_poly)

        return common, result_poly
```

I can see that this is not very efficient. 

_Is there a better way ?_

python-flint can hopefully in the future expose methods like `fmpq_mpoly_get_denominator` or other things I guess but as of now I think we might have to make do with this, maybe this is why we should split this instead of complete override since it's sharing some code right now between both pure python and flint versions.

oscarbenjamin commented:
I don't think that there is a better way to handle `clear_denoms` right now apart from just using `mul_ground`. It can still be useful to separate the implementations though so that we can see in the Flint class where the inefficient things are since we know that we would want to rewrite it in future.

aasim-maverick commented:
> I don't think that there is a better way to handle `clear_denoms` right now apart from just using `mul_ground`. It can still be useful to separate the implementations though so that we can see in the Flint class where the inefficient things are since we know that we would want to rewrite it in future.

Yes, I have kept it under the "divider" for my next commit to separate the implementations.

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3041,50 +2054,1014 @@ def subresultants(self, g, x=None):
+            del p[zm]
+        return p
+
+    def _negate(self):
+        p = self.ring.zero
+        for expv in self.itermonoms():
+            p[expv] = -self._get_coeff(expv)
+        return p
+
+    def _rsub_ground(self, n):
+        p = self._negate()
+        return p._add_ground(n)
```
I haven't done any timings but these both seem inefficient to me. The existing `__neg__` method would likely be faster than this `_negate` by using a list comprehension and initialising a dict once.

The `rsub_ground` here creates a dict with negate and then another new dict with `add_ground`. The existing code in `__rsub__` seems better.

I think it is better not to change the implementation of things like this unless there is some reason and I don't think there is any reason to change this. In the case of FLINT the implementations here would be different anyway like:
```python
def _negate(self):
    return self._from_rep(-self._rep)

def _rsub_ground(self, n):
    return self._from_rep(n - self._rep)
```

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -1100,61 +1021,55 @@ def __rsub__(p1, n):
-        """Multiply two polynomials.
+        ring = self.ring
+
+        if not n:
+            if self:
+                return ring.one
+            else:
+                raise ValueError("0**0")
+        elif len(self) == 1:
+            return self._pow_single_term(n)
+
+        return self._pow_int(n)
```
I don't think that the code before `_pow_int` here makes sense. The subclass can implement these things. The FLINT implementation here would just be:
```python
def _pow_int(self, n):
    return self._from_rep(self._rep ** n)
```
There would be no reason to have a separate `_pow_single_term` method.

oscarbenjamin replied:
Of course the Python subclass can have a `_pow_single_term` method although I'm not sure what the advantage would be of having a method rather than it just being something that is handled inline in `_pow_int`.

aasim-maverick commented on sympy/polys/rings.py:
```diff
@@ -3094,3 +3071,114 @@ def sqf_list(f, all=False):
+        f -= product
+        return symmetric, f
+
+#<-------------------------REFACTORED ABOVE/ TO REFACTOR BELOW-------------------------->#
+
+    def __mul__(p1, p2):
+        """Multiply two polynomials.
```
I think we should either just completely override these functions to have separate implementations for the flint version specifically for the flint supported domains. That way we can avoid unnecessary splits, leverage flint's native functions and avoid complications, 
or we can split these like we did for `__add__` delegating to functions like `_divmod` or `_divmod_ground`(not sure about this)

aasim-maverick committed:
Re-refactored PolyElement (Needs cleaner, more intuitive rearrangment)

aasim-maverick committed:
Corrected locally uncaught quality failure

aasim-maverick committed:
Corrected locally uncaught quality failure for list usage

aasim-maverick commented:
I don't think I should clutter the commits more to pass quality checks, I will look out for similar ones the next time and push. I don't know why but running the quality check locally:

```bash
./bin/test quality
flake8 sympy/
```

gives 0 errors so I thought things are up to the mark.

aasim-maverick committed:
corrected ruff check failure

aasim-maverick commented:
I ran some timings too for the _old v/s new_ implementations of `subs` and `evaluate`, I have used a simple script:

```python
import time
import random
from sympy import ZZ
from sympy.polys.rings import ring


def create_polynomial(num_vars, degree, num_terms):
    var_names = ','.join([chr(ord('x') + i) for i in range(num_vars)])
    R, *variables = ring(var_names, ZZ)

    poly = R(0)
    for _ in range(num_terms):
        coeff = random.randint(1, 20)

        term = R(coeff)
        for var in variables:
            exp = random.randint(0, degree)
            if exp > 0:
                term *= var ** exp

        poly += term

    return poly, variables


def time_method(method, *args):
    start = time.perf_counter()
    result = method(*args)
    end = time.perf_counter()
    return end - start


def benchmark():
    print("PolyElement Benchmark")
    print("=" * 50)

    # format of test cases (num_vars, degree, num_terms)
    test_cases = [
        (1, 5, 10), 
        (2, 5, 20),  
        (3, 5, 30), 
        (4, 5, 40), 
        (2, 10, 50), 
        (3, 10, 100), 
        (5, 5, 100), 
        (4, 8, 200),  
        (6, 6, 300),
    ]

    random.seed(42) 

    for num_vars, degree, num_terms in test_cases:
        print(f"\nTest: {num_vars} vars, degree {degree}, {num_terms} terms")

        poly, variables = create_polynomial(num_vars, degree, num_terms)
        print(f"  Actual terms: {len(poly.terms())}")

        values = [random.randint(1, 5) for _ in variables]
        subs_list = list(zip(variables, values))

        eval_time = time_method(poly.evaluate, subs_list)
        print(f"  evaluate(): {eval_time:.6f}s")

        subs_time = time_method(poly.subs, subs_list)
        print(f"  subs():     {subs_time:.6f}s")

        if len(variables) > 1:
            partial_subs = [(variables[0], values[0])]
            partial_time = time_method(poly.subs, partial_subs)
            print(f"  subs(1var): {partial_time:.6f}s")


if __name__ == "__main__":
    benchmark()
```

Ran this once with the old implementation and once with the new one, haven't used `timeit` here since internally things are cached in SymPy and that has been pointed out to me in the past. 

---

I plotted the results from this rather simple script:

![image](https://github.com/user-attachments/assets/9b687498-28e4-4c1c-9683-288c2a52514e)

The new implementation is faster. 

Shall we keep it ? Or does the new code require changes ?

oscarbenjamin commented:
The timings look good.

aasim-maverick commented:
> Should it not be:
> 
> ```python
>    if len(eval_dict) == ring.ngens:
>         return self._evaluate(eval_dict)
>     else:
>         p = self._subs(eval_dict)
>         ...
> ```

Okay, so computing partial result first and then adjusting `p` to the new ring; Is the following okay ?:

```python
           if len(eval_dict) == ring.ngens:
            return self._evaluate(eval_dict)
        else:
            # compute partial result first then adjust
            p = self._subs(eval_dict)
            new_ring = ring.drop(*[ring.gens[i] for i in eval_dict.keys()])
            result_poly = new_ring.zero

            for monom, coeff in p.iterterms():
                remaining_monom = []
                for i, exp in enumerate(monom):
                    if i not in eval_dict:
                        remaining_monom.append(exp)

                if coeff:
                    remaining_monom = tuple(remaining_monom)
                    if remaining_monom in result_poly:
                        result_poly[remaining_monom] += coeff
                        if not result_poly[remaining_monom]:
                            del result_poly[remaining_monom]
                    else:
                        result_poly[remaining_monom] = coeff

            return result_poly
```

oscarbenjamin commented:
There are functions for converting polynomials between rings already. That code does not need to be reimplemented.

aasim-maverick commented:
> There are functions for converting polynomials between rings already. That code does not need to be reimplemented.

I'm not sure what other way could be used here. 

Also, something I don't really understand is that now when I'm running the same benchmarking script with the old and new implementations, the new one isn't showing the speedups that it showed before(as can be seen in the plots above), I really have no idea why is that happening but this behavior is now persistent across multiple runs and I'm just using this:

```python
    def evaluate(self, *args, **kwargs):
        eval_dict = {}
        ring = self.ring

        if len(args) == 1 and isinstance(args[0], list) and not kwargs:
            for gen, val in args[0]:
                idx = ring.index(gen)
                eval_dict[idx] = ring.domain.convert(val)

        elif len(args) == 2 and not kwargs:
            x, a = args
            idx = ring.index(x)
            eval_dict[idx] = ring.domain.convert(a)

        elif not args and kwargs:
            for gen, val in kwargs.items():
                idx = ring.index(gen)
                eval_dict[idx] = ring.domain.convert(val)
        else:
            raise ValueError("Invalid arguments for evaluate()")

        if not eval_dict:
            return self

        if len(eval_dict) == ring.ngens:
            return self._evaluate(eval_dict)
        else:
            temp_result = self._subs(eval_dict)
            new_ring = ring.drop(*[ring.gens[i] for i in eval_dict.keys()])

            result_poly = new_ring.zero
            remaining_indices = [i for i in range(ring.ngens) if i not in eval_dict]

            for monom, coeff in temp_result.iterterms():
                new_monom = tuple(monom[i] for i in remaining_indices)

                if new_monom in result_poly:
                    result_poly[new_monom] += coeff
                    if not result_poly[new_monom]:
                        del result_poly[new_monom]
                else:
                    if coeff:
                        result_poly[new_monom] = coeff

            return result_poly

    def _evaluate(self, eval_dict):
        result = self.ring.domain.zero

        for monom, coeff in self.iterterms():
            monom_value = self.ring.domain.one
            for i, exp in enumerate(monom):
                if exp > 0:
                    monom_value *= eval_dict[i] ** exp

            result += coeff * monom_value

        return result

    def subs(self, *args, **kwargs):
        subs_dict = {}
        ring = self.ring

        if len(args) == 1 and isinstance(args[0], list) and not kwargs:
            for gen, val in args[0]:
                idx = ring.index(gen)
                subs_dict[idx] = ring.domain.convert(val)

        elif len(args) == 2 and not kwargs:
            x, a = args
            idx = ring.index(x)
            subs_dict[idx] = ring.domain.convert(a)

        elif not args and kwargs:
            for gen, val in kwargs.items():
                idx = ring.index(gen)
                subs_dict[idx] = ring.domain.convert(val)
        else:
            raise ValueError("Invalid arguments for subs()")

        if not subs_dict:
            return self

        if len(subs_dict) == ring.ngens:
            result = self._evaluate(subs_dict)
            return ring.ground_new(result)
        else:
            return self._subs(subs_dict)

    def _subs(self, subs_dict):
        ring = self.ring
        result_poly = ring.zero

        for monom, coeff in self.iterterms():
            new_coeff = coeff
            new_monom = list(monom)

            for i, val in subs_dict.items():
                exp = monom[i]
                if exp > 0:
                    new_coeff *= val ** exp
                new_monom[i] = 0

            if new_coeff:
                new_monom = tuple(new_monom)
                if new_monom in result_poly:
                    result_poly[new_monom] += new_coeff
                    if not result_poly[new_monom]:
                        del result_poly[new_monom]
                else:
                    result_poly[new_monom] = new_coeff

        return result_poly
``` 

Unfortunately if this behavior persists, then I'm not sure if it's really beneficial to switch the old implementations with the new one.

aasim-maverick committed:
__rsub__ reverted back to original split
symmetrize moved in the subclass region
evaluate and subs refactored(but might need reverting ultimately)

aasim-maverick committed:
quality fix

aasim-maverick commented:
I think we will be able to conclude with this PR soon. 

Nothing much has been changed in the last two commits, `__rsub__`'s implementation has been reverted back to the original, `symmetrize` has been moved with a comment, `evaluate` and `subs` have new implementations, but we might revert back to the originals anyways since for some reasons the speedups they showcased weren't consistent after a few runs. 

The arrangement of methods here isn't the most intuitive one in general but it makes it clear what will go in the subclass and what will remain in the base-class which is really useful given how huge the class `PolyElement` is. 
In my opinion, normally the most intuitive way to arrange the methods is to group them according to their purpose, eg. `arithmetic`, `utility`, `properties` etc. and place the helpers next to the methods they aid instead of way down in the class far from where they are actually used. But again, this set up clears what will be overridden and what won't be. 

If preferred, then before merging I will rearrange the methods in the more intuitive way leaving some indicative comments on the methods that will be overridden. Anyways, I locally am aware now which methods will be overridden. 

Also, the test coverage for `rings.py` isn't 100% yet, so if this PR requires it, then that can be done here too.

oscarbenjamin commented:
> > There are functions for converting polynomials between rings already. That code does not need to be reimplemented.
> 
> I'm not sure what other way could be used here.

There is e.g. `set_ring`.

oscarbenjamin commented:
> `evaluate` and `subs` have new implementations, but we might revert back to the originals anyways since for some reasons the speedups they showcased weren't consistent after a few runs.

The main reason for rearranging these methods is to make them more well defined and properly typed. We don't want them to slow down so it is important to do timings if the implementation has changed but the rearrangement is still needed regardless.

aasim-maverick commented:
> > > There are functions for converting polynomials between rings already. That code does not need to be reimplemented.
> > 
> > 
> > I'm not sure what other way could be used here.
> 
> There is e.g. `set_ring`.

Yes that's what I thought of once too, but then I thought maybe `set_ring` won't do the "projection" the method needs here, but yes it should work like this I think:

```python
        if len(eval_dict) == ring.ngens:
            return self._evaluate(eval_dict)
        else:
            temp_result = self._subs(eval_dict)
            new_ring = ring.drop(*[ring.gens[i] for i in eval_dict.keys()])
            return temp_result.set_ring(new_ring)
            ...
```

Thank you.

aasim-maverick committed:
Refactor partial evaluation logic in evaluate

oscarbenjamin commented on sympy/polys/rings.py:
```diff
@@ -3041,6 +3032,156 @@ def subresultants(self, g, x=None):
+
+        return result_poly
+
+    def symmetrize(self):
+        r"""
+        Rewrite *self* in terms of elementary symmetric polynomials.
+
+        Explanation
+        ===========
```
With a method like this that has a large docstring it would be better to keep the method and docstring in the superclass and then have it call `_symmetrize` which would be the subclass implementation.

aasim-maverick committed:
Refactored methods with large docstrings
ruff formatting done

aasim-maverick committed:
Moved helpers to their logical positions

oscarbenjamin commented:
> I think this is ready for review now, before the final cleanup, anymore plausible changes and the test coverage expansion I guess.

I'm not sure if you are waiting for some review here or not. I don't really want to go through the whole diff several times so if it you are still working on things like test coverage or other changes then it is best to finish that and then ask for review.

oscarbenjamin commented:
One way to check this sort of thing is to put `1 / 0` or `assert False` there and then run the whole test suite to see if anything fails. In this case though it seems clear that the line is unreachable.

aasim-maverick commented:
> > So `if not qv` will not trigger and this branch will never be entered.
> 
> I think that you are correct but I would just not worry about covering that line rather than changing the code in this PR.

I agree. Maybe this can be another small patch for later apart from this PR.

aasim-maverick commented:
Currently, locally on my side, the test coverage is at 99%, I am not sure if it should necessarily touch the 100% mark since somethings don't have proper cases to tests with, maybe we don't even need those branches that aren't hit in any natural test case.  I'm not sure and am not changing the code in this PR at least. 

For example though,
```python
    def quo_ground(f, x):
        domain = f.ring.domain

        if not x:
            raise ZeroDivisionError("polynomial division")
        if not f or x == domain.one:
            return f
        return f._quo_ground(x)
```

This function has 6 usages in total, and if I'm not wrong then in none of those places, a zero polynomial seems to be passed as input to this function, and hence the error is never raised for any test case(the rest of the function is covered though). One thing that can be done is to create a separate testing function for this `test_PolyElement_quo_ground` and then intentionally use a test case for covering the whole function.

oscarbenjamin commented:
The `quo_ground` method can be called directly so it can certainly be called with zero as an argument. There should be some test for that.

aasim-maverick committed:
test coverage expanded, minor refactors

oscarbenjamin commented on sympy/polys/tests/test_rings.py:
```diff
@@ -182,10 +336,32 @@ def test_sring():
     R, x, y, z = ring("x,y,z", QQ)
     assert hash(x*y*z)
 
+def test_PolyElement___getnewargs__():
+    R, x, y = ring("x,y", QQ, lex)
+    poly = x**2 + 2*x * y + 3
+    args = poly.__getnewargs__()
```
We don't generally test things like `__getnewargs__` directly. The purpose is to make pickle work and so that it what should be tested:
```python
In [1]:     R, x, y = ring("x,y", QQ, lex)
   ...:     poly = x**2 + 2*x * y + 3

In [2]: import pickle

In [3]: pickle.dumps(poly)
---------------------------------------------------------------------------
RuntimeError
```
If you delete `PolyRing.__getstate__` then it would work:
```python
In [4]: pickle.loads(pickle.dumps(poly))
Out[4]: x**2 + 2*x*y + 3

In [5]: pickle.loads(pickle.dumps(poly)) == poly
Out[5]: True
```
Save that for a different PR though. Here we can just have an XFAIL test for the pickling.

aasim-maverick replied:
> We don't generally test things like `__getnewargs__` directly. The purpose is to make pickle work and so that it what should be tested:
> 
> ```python
> In [1]:     R, x, y = ring("x,y", QQ, lex)
>    ...:     poly = x**2 + 2*x * y + 3
> 
> In [2]: import pickle
> 
> In [3]: pickle.dumps(poly)
> ---------------------------------------------------------------------------
> RuntimeError
> ```
> 
> If you delete `PolyRing.__getstate__` then it would work:
> 
I'm sorry. I think pickling works still, I mean to say it works on my side:

```python
In [1]: R, x, y = ring("x,y", QQ)

In [2]: poly = x**2 + 2*x*y + 3

In [3]: import pickle

In [4]: dumps = pickle.dumps(poly)

In [5]: type(dumps), len(dumps)
Out[5]: (bytes, 654)

In [6]: pickle.loads(dumps)
Out[6]: x**2 + 2*x*y + 3

```

There can be some subtle reasons for this difference on our respective sides, nothing that I'd be too sure of.

oscarbenjamin replied:
It might depend which python version you are using. This is with 3.13:
```
In [1]:     R, x, y = ring("x,y", QQ, lex)
   ...:     poly = x**2 + 2*x * y + 3

In [2]: import pickle

In [3]: pickle.loads(pickle.dumps(poly))
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[3], line 1
----> 1 pickle.loads(pickle.dumps(poly))

File ~/sympy/polys/rings.py:285, in PolyRing.__getstate__(self)
    282 state = self.__dict__.copy()
    283 del state["leading_expv"]
--> 285 for key in state:
    286     if key.startswith("monomial_"):
    287         del state[key]

RuntimeError: dictionary changed size during iteration
```

aasim-maverick replied:
Actually no, now I remember, `__getstate__` was refactored.

aasim-maverick replied:
>Are there other places where the behaviour is changed?

There weren't many such places, most of our attention has been towards `PolyElement` but these are the things that I can remember that I changed in `PolyRing`:

### **a) Generator index computation logic for negative indices in `index`**:
Orignal:
```python
elif -self.ngens <= i and i <= -1:
    i = -i - 1
```
I changed this to:
```python
elif -self.ngens <= gen <= -1:
    return gen + self.ngens
```
No behavioral change because of this, I just think this formula is more intuitive, the way `index` handles negative indices right now is through a rather awkward formula. 

### **b) Dictionary processing logic in `from_dict`**:
Original:
```python
def from_dict(self, element, orig_domain=None):
    domain_new = self.domain_new
    poly = self.zero

    for monom, coeff in element.items():
        coeff = domain_new(coeff, orig_domain)
        if coeff:
            poly[monom] = coeff

    return poly
```
I split it like this:
```python
def from_dict(self, element, orig_domain=None):
    if not isinstance(element, dict):
        raise TypeError("Input must be a dictionary mapping monomials to coefficients")
    return self._from_dict_ground(element, orig_domain)

def _from_dict_ground(self, element, orig_domain=None):
    # ... similar logic but with reordered condition check
    for monom, coeff in element.items():
        if coeff:  # Check coeff first
            coeff = domain_new(coeff, orig_domain)
            poly[monom] = coeff
```
Besides change in order of condition check, also added input validation. No behavioral change because of this.

### **c) `__getstate__`**:
Original:
```python
def __getstate__(self):
    state = self.__dict__.copy()
    del state["leading_expv"]

    for key in state:
        if key.startswith("monomial_"):
            del state[key]

    return state
```
I changed it into something I thought would be safer:
```python
def __getstate__(self):
    state = self.__dict__.copy()

    state.pop("leading_expv", None)

    for key in list(state.keys()):
        if key.startswith("monomial_"):
            state.pop(key, None)

    return state
```
Behavioral change seen here since now this avails pickling which is a desirable change but not this way and also not in this PR.

aasim-maverick committed:
Parameter consistenecy in rings.py
dunder methods tested by dynamic calling
pickle and serialization support tested conventionally
cleanup
quality fix

aasim-maverick committed:
Removed unnecessary classes and their usages from tests

aasim-maverick committed:
Reverted back handling of negative indices in PolyRing.index
Reverted back to the original PolyRing.__getstate__()

aasim-maverick commented:
gh-28187 created a conflict if I'm not wrong.

aasim-maverick committed:
conflict resolution with PR-28187(try1)

aasim-maverick committed:
Merge branch 'master' into rings-rearrangement

aasim-maverick commented:
I was considering organizing the `PolyElement` class similarly to how I did for `PolyRing`, by grouping methods based on their purpose, like arithmetic, properties, utility, etc. However, the current method order in this PR already feels quite reasonable. We've achieved a clearer separation between interface and implementation via our splits and rearrangement, in fact substitution and evaluation are 2-3x faster, and applying the same kind of compartmentalization to `PolyElement` doesn’t seem as effective for now in this PR, mainly because it’s significantly larger and more complex than PolyRing.

I believe a more intuitive grouping of methods for such a large class will be more meaningful once we move to the next stage of the project i.e. setting up the baseclass-subclass hierarchy. I guess we’ll need to revisit and refine the structure during that process anyway, so it makes more sense to defer the rearrangement until then. Although that's just my opinion.

That said, this is ready for review again. I'm happy to make further changes if needed, but overall I guess we might be close to wrapping up this step.

aasim-maverick commented:
I'm not sure, should I make any expected changes before tomorrow's meet ?

oscarbenjamin commented:
I think this looks good. I need to give it a closer look.

We can talk about where to go from here.

oscarbenjamin commented on sympy/polys/tests/test_rings.py:
```diff
@@ -243,6 +436,22 @@ def test_PolyElement__str__():
         assert str(2*t**2 + 1) == 'EX(2)*t**2 + EX(1)'
 
+    R, x = ring("x", ZZ)
+    f = R.zero
+    s = str(f)
+    assert s == "0"
```
Why not just `assert str(R.zero) == 0`?

oscarbenjamin commented on sympy/polys/tests/test_rings.py:
```diff
@@ -813,6 +1084,165 @@ def test_PolyElement___pow__():
+    z_poly = z ** 2 + 1
+    raises(TypeError, lambda: p1 % z_poly)
+
+    r = p1 % 2
+    assert r == x ** 2 + 1
```
Don't need the variable here.

oscarbenjamin commented:
Looks good. Let's get this in. Followup PRs can improve on the test formatting or other things.
